src/names/name-resolution.md

```diff
@@ -1,4 +1,420 @@
+```
+
+Multiple glob imports are allowed to import the same name, and that name is allowed to be used if the imports are of the same item (following re-exports). The visibility of the name is the maximum visibility of the imports. For example:
+
+```rust
+mod foo {
+    pub struct Qux;
+}
+
+mod bar {
+    pub use super::foo::Qux;
+}
+
+// These both import the same `Qux`. The visibility of `Qux`
+// is `pub` because that is the maximum visibility between
+// these two `use` declarations.
+pub use bar::*;
+use foo::*;
+
+fn main() {
+    let _: Qux = Qux;
+}
+```
+
+r[names.resolution.expansion.imports.ambiguity.glob-vs-outer]
+Names may not be resolved through glob imports when there is another candidate available in an [outer scope].
+
+```rust,compile_fail,E0659
+mod bar {
+    pub mod foo {
+//          ^-- glob `foo` candidate
+        pub struct Name;
+    }
+}
+
+pub mod foo {
+//      ^-- outer `foo` candidate
+    pub struct Name;
+}
+
+pub fn qux() {
+    use bar::*;
+    use foo::Name; // ERROR: `foo` is ambiguous
+}
+```
+
+```rust,compile_fail,E0659
+pub mod bar {
+    #[macro_export]
+    macro_rules! m {
+        () => {};
+    }
+
+    macro_rules! m2 {
+        () => {};
+    }
+    pub(crate) use m2 as m;
+}
+
+pub fn qux() {
+    use bar::*;
+    m!(); // ERROR: `m` is ambiguous
+}
+```
+
+> [!NOTE]
+> These ambiguity errors are specific to imports, even though they are only observed when those imports are used. Having multiple candidates available for a given name during later stages of resolution is not considered an error. So long as none of the imports themselves are ambiguous, there will always be a single unambiguous closest resolution.
+>
+> ```rust
+> mod bar {
+>     pub const NAME: bool = true;
+> }
+>
+> mod baz {
+>     pub const NAME: bool = false;
+> }
+>
+> use baz::NAME;
+>
+> pub fn foo() {
+>     use bar::*;
+>     assert!(NAME);
+>     //      ^--- This `NAME` is resolved during primary resolution.
+> }
+> ```
+
+r[names.resolution.expansion.imports.ambiguity.path-vs-textual-macro]
+Path-based scope bindings for macros may not shadow textual scope bindings to macros.
+
+```rust,compile_fail,E0659
+macro_rules! m2 {
+    () => {}
+}
+macro_rules! m {
+    () => {}
+}
+pub fn foo() {
+    m!(); // ERROR: `m` is ambiguous
+    use m2 as m; // In scope for entire function body.
+}
+```
+
+r[names.resolution.expansion.macros]
+### Macros
+
+r[names.resolution.expansion.macros.intro]
+Macros are resolved by iterating through the available scopes to find the available candidates. Macros are split into two sub-namespaces, one for bang macros, and the other for attributes and derives. Resolution candidates from the incorrect sub-namespace are ignored.
+
+r[names.resolution.expansion.macros.visitation-order]
+The available scopes are visited in the following order.
+
+* [Derive helpers]
+* [Textual scope macros]
+* [Path-based scope macros]
+* [`macro_use` prelude]
+* [Standard library prelude]
+* [Builtin attributes]
```

Comment:
should this clarify that each of these bullets represent one or more scopes or is this intuitively obvious?